### ISSUE TITLE
[BUG] JSQLParser Version : RDBMS : JSQLParserException: Time out occurred.

### ISSUE THREAD

**Issue by young0098** [JSQLParser/JSqlParser#1897]
Always check against the **Latest SNAPSHOT of JSQLParser** and the [Syntax Diagram](https://jsqlparser.github.io/JSqlParser/syntax_snapshot.html)

### Failing SQL Feature:
- parse failed because of time-out error
- net.sf.jsqlparser.JSQLParserException: Time out occurred.
### SQL Example:
- Simplified Query Example, focusing on the failing feature
    ```sql
    INSERT INTO TMP_CAL_FST_SETL_DATE(
	MULTI_TENANCY_ID 
	, LOAN_ACCNO 
	, LNAR_ID 
	, CAL_FST_SETL_DATE 
    )
    SELECT
	P1.MULTI_TENANCY_ID AS MULTI_TENANCY_ID
	,P1.CST_ACCNO AS LOAN_ACCNO
	,P1.LNAR_ID AS LNAR_ID
	, CASE WHEN P1.LN_FTM_INTSETL_DT>TO_DATE('1900-01-02','YYYY-MM-DD') AND P1.LN_FTM_INTSETL_DT<TO_DATE('2099-12-31','YYYY-MM-DD') THEN P1.LN_FTM_INTSETL_DT::DATE
	WHEN P2.INTSETL_MTDCD='01' THEN CASE WHEN P1.FST_DSBR_DT='' THEN TO_DATE('1900-01-01','YYYY-MM-DD')::DATE ELSE P1.FST_DSBR_DT::DATE END
	WHEN P2.INTSETL_MTDCD='03' THEN (CASE WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)NOT IN('DD','')
	THEN (CASE WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)=SUBSTR(P1.FST_DSBR_DT::DATE,9,2)
	THEN (CASE WHEN INTSETL_CYC_ECD IS NOT NULL AND SUBSTR(INTSETL_CYC_ECD,2,1)>0 THEN ADD_MONTHS(P1.FST_DSBR_DT::DATE,CAST(SUBSTR(INTSETL_CYC_ECD,2,2)AS INTEGER))::DATE
	WHEN INTSETL_CYC_ECD IS NOT NULL AND SUBSTR(INTSETL_CYC_ECD,2,1)=0 THEN ADD_MONTHS(P1.FST_DSBR_DT::DATE,CAST(SUBSTR(INTSETL_CYC_ECD,3,1)AS INTEGER))
	END
	)::DATE
	WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)<SUBSTR(P1.FST_DSBR_DT::DATE,9,2)
	THEN (CASE WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)<=SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,2),'MM')-1,9,2) THEN DATE(SUBSTR(P1.FST_DSBR_DT::DATE,1,8)||SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2))::DATE
	WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)>SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,2),'MM')-1,9,2) THEN DATE(SUBSTR(P1.FST_DSBR_DT::DATE,1,8)||SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,2),'MM')-1,9,2)) END )::DATE
	WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)>SUBSTR(P1.FST_DSBR_DT::DATE,9,2)
	THEN (CASE WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)<=SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,1),'MM')-1,9,2) THEN DATE(SUBSTR(P1.FST_DSBR_DT::DATE,1,8)||SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2))::DATE
	WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)>SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,1),'MM')-1,9,2) THEN TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,2),'MM')-1 END )
	END )
	WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)='DD'
	THEN (CASE WHEN INTSETL_CYC_ECD IS NOT NULL AND SUBSTR(INTSETL_CYC_ECD,2,1)>0 THEN ADD_MONTHS(P1.FST_DSBR_DT::DATE,CAST(SUBSTR(INTSETL_CYC_ECD,2,2) AS INTEGER))::DATE
	WHEN INTSETL_CYC_ECD IS NOT NULL AND SUBSTR(INTSETL_CYC_ECD,2,1)=0 THEN ADD_MONTHS(P1.FST_DSBR_DT::DATE,CAST(SUBSTR(INTSETL_CYC_ECD,3,1) AS INTEGER))::DATE
	ELSE P1.FST_DSBR_DT::DATE END )::DATE END )::DATE
	ELSE TO_DATE('1900-01-01','YYYY-MM-DD') END ::DATE AS CAL_FST_SETL_DATE
FROM ${V_BASE}.TBCLLAC0_H P1 
LEFT JOIN ${V_BASE}.TBCLLAR0_H P2 
	ON P2.START_DATE<=TO_DATE('${TXNDATE}','YYYY-MM-DD')
AND P2.END_DATE>TO_DATE('${TXNDATE}','YYYY-MM-DD')
AND P1.MULTI_TENANCY_ID=P2.MULTI_TENANCY_ID
AND P1.LNAR_ID=P2.LNAR_ID AND SUBSTR(P2.SPLIT_BRANCH_CD,1,5)='${MTID}'
WHERE P1.START_DATE<=TO_DATE('${TXNDATE}','YYYY-MM-DD')
AND P1.END_DATE>TO_DATE('${TXNDATE}','YYYY-MM-DD') AND SUBSTR(P1.SPLIT_BRANCH_CD,1,5)='${MTID}'
    ```

### Software Information:
- JSqlParser 4.6
- PostgreSQL

### Tips:
Please write in English and avoid Screenshots (as we can't copy and paste content from it).
[Try your example online with the latest JSQLParser](http://jsqlformatter.manticore-projects.com) and share the link in the error report.
Do provide Links or References to the specific Grammar and Syntax you are trying to use.

**Comment by manticore-projects** [comment]
You can increase the timeout as shown [here](https://manticore-projects.com/JSQLParser/usage.html#define-the-parser-features).

**Comment by young0098** [comment]
Thanks，I set parser timeout to 60000 ms, It seems can be parsed continued, but now encounter new error:net.sf.jsqlparser.JSQLParserException: net.sf.jsqlparser.parser.ParseException: Encountered unexpected token: "::" "::"
    at line 29, column 45.

**Comment by manticore-projects** [comment]
It means that your statement is somewhere syntactically wrong.
If you want to solve this, then maybe format your SQL properly so someone can follow the logic behind.

**Comment by manticore-projects** [comment]
`FROM ${V_BASE}` this will definitely not work.

**Comment by manticore-projects** [comment]
This seems to work:
```sql
INSERT INTO tmp_cal_fst_setl_date (
    multi_tenancy_id
    , loan_accno
    , lnar_id
    , cal_fst_setl_date )
SELECT  p1.multi_tenancy_id AS multi_tenancy_id
        , p1.cst_accno AS loan_accno
        , p1.lnar_id AS lnar_id
        , CASE 
            WHEN P1.LN_FTM_INTSETL_DT > TO_DATE('1900-01-02', 'YYYY-MM-DD') AND P1.LN_FTM_INTSETL_DT < TO_DATE('2099-12-31', 'YYYY-MM-DD') 
            THEN P1.LN_FTM_INTSETL_DT::DATE 
            WHEN P2.INTSETL_MTDCD = '01' 
            THEN CASE 
                    WHEN P1.FST_DSBR_DT = '' 
                    THEN TO_DATE('1900-01-01', 'YYYY-MM-DD')::DATE 
                    ELSE P1.FST_DSBR_DT::DATE END WHEN P2.INTSETL_MTDCD = '03' 
                    THEN (CASE 
                            WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)NOT IN('DD','')
                            THEN (CASE 
                                        WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)=SUBSTR(P1.FST_DSBR_DT::DATE,9,2)
                                        THEN (CASE 
                                                    WHEN INTSETL_CYC_ECD IS NOT NULL AND SUBSTR(INTSETL_CYC_ECD,2,1)>0 
                                                    THEN ADD_MONTHS(P1.FST_DSBR_DT::DATE,CAST(SUBSTR(INTSETL_CYC_ECD,2,2)AS INTEGER))::DATE
                                                    WHEN INTSETL_CYC_ECD IS NOT NULL AND SUBSTR(INTSETL_CYC_ECD,2,1)=0 
                                                    THEN ADD_MONTHS(P1.FST_DSBR_DT::DATE,CAST(SUBSTR(INTSETL_CYC_ECD,3,1)AS INTEGER))
                                                    END)::DATE
                                        WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)<SUBSTR(P1.FST_DSBR_DT::DATE,9,2)
                                        THEN (CASE 
                                                    WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)<=SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,2),'MM')-1,9,2) 
                                                    THEN DATE(SUBSTR(P1.FST_DSBR_DT::DATE,1,8)||SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2))::DATE
                                                    WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)>SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,2),'MM')-1,9,2) 
                                                    THEN DATE(SUBSTR(P1.FST_DSBR_DT::DATE,1,8)||SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,2),'MM')-1,9,2)) 
                                                    END )::DATE
                                        WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)>SUBSTR(P1.FST_DSBR_DT::DATE,9,2)
                                        THEN (CASE 
                                                    WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)<=SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,1),'MM')-1,9,2) 
                                                    THEN DATE(SUBSTR(P1.FST_DSBR_DT::DATE,1,8)||SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2))::DATE
                                                    WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)>SUBSTR(TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,1),'MM')-1,9,2) 
                                                    THEN TRUNC(ADD_MONTHS(P1.FST_DSBR_DT::DATE,2),'MM')-1 
                                                    END )
                                    END )
                            WHEN SUBSTR(P2.APT_INTSETL_DAY_DEF,5,2)='DD'
                            THEN (CASE WHEN INTSETL_CYC_ECD IS NOT NULL AND SUBSTR(INTSETL_CYC_ECD,2,1)>0 THEN ADD_MONTHS(P1.FST_DSBR_DT::DATE,CAST(SUBSTR(INTSETL_CYC_ECD,2,2) AS INTEGER))::DATE
                                    WHEN INTSETL_CYC_ECD IS NOT NULL AND SUBSTR(INTSETL_CYC_ECD,2,1)=0 THEN ADD_MONTHS(P1.FST_DSBR_DT::DATE,CAST(SUBSTR(INTSETL_CYC_ECD,3,1) AS INTEGER))::DATE
                                    ELSE P1.FST_DSBR_DT::DATE END )::DATE 
                            END )::DATE
            ELSE TO_DATE('1900-01-01', 'YYYY-MM-DD') 
            END::DATE AS cal_fst_setl_date
FROM /*${V_BASE}*/ base_schema.tbcllac0_h p1
    LEFT JOIN /*${V_BASE}*/ base_schema.tbcllar0_h p2
        ON p2.start_date <= To_Date( '${TXNDATE}', 'YYYY-MM-DD' )
            AND p2.end_date > To_Date( '${TXNDATE}', 'YYYY-MM-DD' )
            AND p1.multi_tenancy_id = p2.multi_tenancy_id
            AND p1.lnar_id = p2.lnar_id
            AND Substr( p2.split_branch_cd, 1, 5 ) = '${MTID}'
WHERE p1.start_date <= To_Date( '${TXNDATE}', 'YYYY-MM-DD' )
    AND p1.end_date > To_Date( '${TXNDATE}', 'YYYY-MM-DD' )
    AND Substr( p1.split_branch_cd, 1, 5 ) = '${MTID}'
;
```

**Comment by young0098** [comment]
but on my laptop，use JSqlParser 4.6，paser your formated sql, still prompt error message: net.sf.jsqlparser.JSQLParserException: net.sf.jsqlparser.parser.ParseException: Encountered unexpected token: "::" "::" at line 46, column 34.

**Comment by manticore-projects** [comment]
The corrected statement works well with JSQLParser 4.8 Snapshot and [you can test it online here.](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20JIOQygogSgKgBKGB5OAXAtgBwPoGMCGANtgGYDOq2ZApqsQCb6rVwAUAUHF3OgK6GoAltmYA7fKNwBPbIPqduAGjiEA9hOz5cuUaoVdlhcQCdZ87nGUFi5SjTrZGzOAEp2kADIQAwvDiYARgA6PgFhMQlpMzgAQTAefiERanFJGTl9C0t-YNwKTW1dWPi1DS0dPSylHKCjfFM5YpUTM0yLZW84iDg2qrgAdQAJCBA4AAVgjxBsADEYAFlsREgYD2wAEXgAPjhkDZiYCFYAcgCATgAGC4BaC4DbgCZj5WOATXfX6-n56-X145csRA63Gk2mc0Wywgqw28AAPLskPtDicHhczmdrgEHtcAMwBZ5wN4fL4-P4Anp9LIwYajCZBKazBZLEAwFZrTYALk56wO3V6VSGI3GDyCUJh8xg628IIAvES7sdKVTuDThZ1IMqVSqhXTgjMwDANmAAEJQWFweXHJUC7Wq2mI5FHU6XG53W4El4fT7fX7-Fzc3mHLV2qoQDya+kGo3rU3mrk8vlwEYg3UisWs9nYSXSuUK3E20NUtWjVga-lFysDB1gACqJsNUFYY1FMTGRvFHJirw2EBmigArIoHi4QEh4KATuTFNa3FWiyW2OWQ-PV9w03WGzAmy2gm2O5noV2e+s+4Phy5ZZvG839YbjWbYYG+YozhfbWvK4uy10V5--1Saadtg3ivN42A+CCwDxGO8AgLWHgeECILXturDAaB4GQcOigBC4WwXH+AHEfawoxH82ZIKygxgLeQTRg+8YwM+hyKBqMCsKhTYYWBEEyjhI5xAgrIQAA4tALgBomhwfiR-5AYeMKYXxUEweOcDwYhyFwFx6GKWsynYQ8uGXoRslyWui7keslHUbRUb3rGj4JkGEBsXEHG6TxWH8biJlCYgYkSXOFmhSmUmueZoXrjW9Y3ru+4smyR77CeZ5DiOcK6Q5MZxk+0luW+I5RdFuwOj+molaVdobnFaEJe2SVZryaX9hlLhwledVNtutYgN4rDWbZNL2XeuXOcxBUXjO3wAlir4XkR1Xzourmcd1dEMU5TEsW5ASKAAHC4AA+x3Za2jXAS1vZtReEV8lVy2AbFW47hdB7JTC12nrdI5bLpvX9YNFHzFRI2bY5eUuS+I4zfMc37UVgKPU9VSrXy62vRD407VN+1HadANQH1A1DaDdnY4x+WudNxyzS481I8jqP-imri7Sjcm1VjDUfc13Y3eef3ZWNVPQ6xSOcyR37LlLLPVsK517pd+mpYL7WdUTJPAzZ5PgzlYuTTTeFwwjC0jkt8vaujKIi-RkMTbtuGHSdZ0bbzTUpd96V3RzVurtz8XvZ7X0Cz9Qv4VrQNk2DNGU9t1MvibdPwwziOLXLLOLoDpMg7Ho32zjiesbDKcI5b-t9GzIXy9XmfPYr7vB1dYc+yOsrHOS9d9DLv4KZ9Bm8ZBCBqXBCFITEwI6Rt3kqTheEEWVZF5xTBsJ+LbnsZjN6z0Zi0BSJ4lQJJftW-3WaGTKI8aepmkT1PXmq5f6zz6ZS+jDHq+i+vRsvlvj8DxAkPXyJkmiBSPifAq3dKzhkjN-KGv9gzV12hXNcyCoGhlgd0PYa0XRXA9B6QkxJ3ikj9BSKKKYUFCWsKQfI9gGBMGoOwGYUAkDzDgAAegAFQABIADeAA1bAJougAF8uEcLgAAI3wDQKguAAAW1B0D4CCKgKRuBCCEC0BcbACiciZC8HMOAAApJAoBOG8MEcIsREjpGyOoPIpRKi1EaK0fUXR+jMAPFtFRfwooKD1EoE4FgnVdiqA2Iw1gRJ+EwAABrAj5KIoh3pSHklcFFSeIJvFBBSPQRwjC4A7BgBE9YUSYl8PiYkw4ySvQkl9OkmuKosk1FCEkCIaRojyhyW08IKRIjpHMHaFpgRagtEaN00UdQGhDO1C0sAvApEUGMNEnJZBMCEEEJQKRxhIgKLwPQZQ+04ADlcJaCpkpgDrGSewIUUBuijMCcYYJhSwklMicwaJxxYkJNcrUokqSGn-AyRYEZwQ8kFOcMU0p5TvmVN+UklJ9SyTAqafMxZyzVnBHWZs7ZuzJD7NwIcuAxzTmAitPwy51zjjsAANxAA)
